File: NodejsBackendKit-ApiKey/PaymentHandler.js
const fs = require("fs");
const https = require("https");
const { EOL } = require("os");
const path = require("path");
const crypto = require("crypto");

class SimpleLogger {
  disableLogging = false;
  constructor(logFilePath) {
    this.disableLogging = false;
    if (SimpleLogger.instance !== undefined) {
      return SimpleLogger.instance;
    }
    const dirname = path.dirname(logFilePath);
    if (!fs.existsSync(dirname)) {
      fs.mkdirSync(dirname, { recursive: true });
    }
    this.logFilePath = logFilePath;
    SimpleLogger.instance = this;
    return SimpleLogger.instance;
  }

  log(level, apiTag, paymentRequestId, message, value) {
    const timestamp = this.formatDateTime(Date.now());
    let valueStr = value;
    if (typeof value === "object") {
      valueStr = JSON.stringify(value);
    }
    const logMessage = `${timestamp} [${level.toUpperCase()}] apiTag=${apiTag}, paymentRequestId=${paymentRequestId}, message=${message}, value=${valueStr}${EOL}`;
    fs.appendFile(this.logFilePath, logMessage, () => {});
  }

  info(apiTag, paymentRequestId, message, value) {
    this.log("info", apiTag, paymentRequestId, message, value);
  }

  error(apiTag, paymentRequestId, message, value) {
    this.log("error", apiTag, paymentRequestId, message, value);
  }

  formatDateTime(timestamp) {
    const date = new Date(timestamp);
    const months = [
      "Jan",
      "Feb",
      "Mar",
      "Apr",
      "May",
      "Jun",
      "Jul",
      "Aug",
      "Sep",
      "Oct",
      "Nov",
      "Dec",
    ];
    const year = date.getFullYear();
    const month = months[date.getMonth()];
    const day = date.getDate();
    let hour = date.getHours();
    const minute = date.getMinutes();
    const second = date.getSeconds();
    const period = hour >= 12 ? "PM" : "AM";
    hour = hour % 12;
    hour = hour ? hour : 12;
    return `${month} ${day}, ${year} ${hour}:${minute}:${second} ${period}`;
  }
}

/**
 * Singleton class
 */
class PaymentHandler {
  /**
   * use this as it facilitates with some basic error handling
   */
  static getInstance(userSpecifiedConfigPath) {
    return new PaymentHandler(userSpecifiedConfigPath);
  }

  constructor(userSpecifiedConfigPath) {
    if (PaymentHandler.paymentHandlerInstance !== undefined)
      return PaymentHandler.paymentHandlerInstance;
    this.paymentConfigs = undefined;
    const configPath = userSpecifiedConfigPath || "config.json";
    try {
      const config = fs.readFileSync(configPath, "utf-8");
      this.paymentConfigs = JSON.parse(config);
    } catch (error) {
      console.error(
        "Failed to read configs from file, here's tbe error message:- " +
          error.message
      );
      throw new TypeError("Failed to find/read config file");
    }
    this.validatePaymentConfigs();
    this.logger = new SimpleLogger(this.getLoggingPath());
    this.logger.disableLogging = !this.getEnableLogging();
    PaymentHandler.paymentHandlerInstance = this;
    return PaymentHandler.paymentHandlerInstance;
  }

  orderSession(params) {
    this.validateParams(params);
    return this.makeServiceCall({
      apiTag: "ORDER_SESSION",
      method: "POST",
      path: "/session",
      body: {
        payment_page_client_id: this.getPaymentPageClientId(),
        ...params,
      },
      headers: {
        "Content-Type": "application/json",
      },
    });
  }

  orderStatus(params) {
    let orderId = undefined;
    if (typeof params === "string") {
      orderId = params;
    } else {
      this.validateParams(params);
      orderId = params["order_id"];
    }
    if (orderId === undefined) {
      throw new APIException(
        -1,
        "INVALID_PARAMS",
        "INVALID_PARAMS",
        "order_id is missing, usage:- orderStatus('order_id') or orderStatus({'order_id': 'value', ...other configs here})"
      );
    }
    return this.makeServiceCall({
      apiTag: "ORDER_STATUS",
      method: "GET",
      path: `/orders/${orderId}`,
      body: params,
    });
  }

  refund(params) {
    this.validateParams(params);
    return this.makeServiceCall({
      apiTag: "ORDER_REFUND",
      method: "POST",
      path: `/refunds`,
      body: params,
    });
  }

  // utility functions
  makeServiceCall({ apiTag, path, method, headers = {}, query = {}, body }) {
    return new Promise((resolve, reject) => {
      const paymentRequestId = this.generateUUID();
      headers = {
        "Content-Type": "application/x-www-form-urlencoded",
        "User-Agent": `NODEJS_KIT/1.0.0`,
        version: this.getVersion(),
        "x-merchantid": this.getMerchantId(),
        ...headers,
        Authorization: "Basic " + this.base64EncodeKey(this.getApiKey()),
      };

      const payload = this.prepareData(body, headers);
      const encodedQueryParams = this.prepareData(query);
      const pathWithQueryParams =
        path +
        (encodedQueryParams.length === 0 ? "" : `?${encodedQueryParams}`);

      this.logger.info(apiTag, paymentRequestId, "Request parameters", payload);

      const fullPath = new URL(pathWithQueryParams, this.getBaseUrl()),
        agent = new https.Agent({ keepAlive: true });

      this.logger.info(
        apiTag,
        paymentRequestId,
        "Executing request",
        fullPath.pathname + fullPath.search
      );

      const httpOptions = {
        host: fullPath.host,
        port: fullPath.port,
        path: fullPath.pathname + fullPath.search,
        method,
        agent,
        headers,
      };

      const req = https.request(httpOptions);

      req.setTimeout(100000, () => {
        this.logger.log(apiTag, paymentRequestId, {
          message: "Request has been timedout",
        });
        req.destroy(
          new APIException(
            -1,
            "REQUEST_TIMEOUT",
            "REQUEST_TIMEOUT",
            "Request has been timedout"
          )
        );
      });

      req.on("response", (res) => {
        res.setEncoding("utf-8");
        this.logger.info(
          apiTag,
          paymentRequestId,
          "Received Http Response Code",
          res.statusCode
        );

        let responseBody = "";
        res.on("data", (chunk) => {
          responseBody += chunk;
        });

        res.once("end", () => {
          this.logger.info(
            apiTag,
            paymentRequestId,
            "Received Response",
            responseBody
          );
          let resJson = undefined;
          try {
            resJson = JSON.parse(responseBody);
          } catch (e) {
            // in sdk we only accept json response
            this.logger.error(
              apiTag,
              paymentRequestId,
              "INVALID_RESPONSE_FORMAT",
              e
            );
            return reject(
              new APIException(
                res.statusCode,
                "INVALID_RESPONSE",
                "INVALID_RESPONSE",
                responseBody || "Failed to parse response json"
              )
            );
          }

          if (res.statusCode >= 200 && res.statusCode < 300) {
            return resolve(resJson);
          } else {
            let status = resJson["status"],
              errorCode = resJson["error_code"],
              errorMessage = resJson["error_message"];
            return reject(
              new APIException(res.statusCode, status, errorCode, errorMessage)
            );
          }
        });
      });

      req.once("socket", (socket) => {
        if (socket.connecting) {
          socket.once("secureConnect", () => {
            req.write(payload);
            req.end();
          });
        } else {
          req.write(payload);
          req.end();
        }
      });

      req.on("error", (error) => {
        this.logger.error(
          apiTag,
          paymentRequestId,
          "Please check your internet connection/Failed to establish connection",
          error
        );
        return reject(error);
      });
    });
  }

  validateParams(params) {
    if (typeof params != "object")
      throw new APIException(
        -1,
        "INVALID_PARAMS",
        "INVALID_PARAMS",
        "Params are empty or non an object"
      );
  }

  validatePaymentConfigs() {
    const nonNullConfigs = [
      "MERCHANT_ID",
      "API_KEY",
      "PAYMENT_PAGE_CLIENT_ID",
      "BASE_URL",
      "ENABLE_LOGGING",
      "LOGGING_PATH",
      "RESPONSE_KEY",
    ];
    for (let key of nonNullConfigs) {
      if (!(key in this.paymentConfigs)) {
        console.error(key + " not present in config.");
        throw new TypeError(key + " not present in config.");
      }
    }
  }

  prepareData(data, headers) {
    if (data === undefined) return "";
    if (typeof data != "object") return String(data);
    if (
      headers !== undefined &&
      headers["Content-Type"] === "application/json"
    ) {
      return JSON.stringify(data);
    }
    return Object.keys(data)
      .map(
        (key) => encodeURIComponent(key) + "=" + encodeURIComponent(data[key])
      )
      .join("&");
  }

  base64EncodeKey(str) {
    return Buffer.from(str).toString("base64");
  }

  generateUUID = () => {
    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(
      /[xy]/g,
      function (c) {
        const r = (Math.random() * 16) | 0;
        const v = c === "x" ? r : (r & 0x3) | 0x8;
        return v.toString(16);
      }
    );
  };

  // helpers getters and setters
  getMerchantId() {
    return this.paymentConfigs.MERCHANT_ID;
  }

  getBaseUrl() {
    return this.paymentConfigs.BASE_URL;
  }

  getApiKey() {
    return this.paymentConfigs.API_KEY;
  }

  getPaymentPageClientId() {
    return this.paymentConfigs.PAYMENT_PAGE_CLIENT_ID;
  }

  getEnableLogging() {
    return typeof this.paymentConfigs.ENABLE_LOGGING == "boolean"
      ? !this.paymentConfigs.ENABLE_LOGGING
      : true;
  }

  getLoggingPath() {
    return (
      this.paymentConfigs.LOGGING_PATH ||
      path.join(__dirname, "logs/paymentHandler.log")
    );
  }

  getResponseKey() {
    return this.paymentConfigs.RESPONSE_KEY;
  }

  getVersion() {
    return this.version || "2024-06-24";
  }
}

class APIException extends Error {
  status = undefined;
  errorCode = undefined;
  errorMessage = undefined;
  httpResponseCode = undefined;

  constructor(httpResponseCode, status, errorCode, errorMessage) {
    super(errorMessage || errorCode || "Something went wrong");
    this.status = status;
    this.errorCode = errorCode;
    this.errorMessage = errorMessage;
    this.httpResponseCode = httpResponseCode;
  }
}

function validateHMAC_SHA256(params, secretKey) {
  if (typeof params !== "object" || typeof secretKey !== "string") {
    throw new TypeError(
      "params should be object, given ",
      typeof params,
      ", secretKey should be string, given ",
      typeof secretKey
    );
  }

  var paramsList = [];
  for (var key in params) {
    if (key != "signature" && key != "signature_algorithm") {
      paramsList[key] = params[key];
    }
  }

  paramsList = sortObjectByKeys(paramsList);

  var paramsString = "";
  for (var key in paramsList) {
    paramsString = paramsString + key + "=" + paramsList[key] + "&";
  }

  let encodedParams = encodeURIComponent(
    paramsString.substring(0, paramsString.length - 1)
  );
  let computedHmac = crypto
    .createHmac("sha256", secretKey)
    .update(encodedParams)
    .digest("base64");
  let receivedHmac = decodeURIComponent(params.signature);

  return decodeURIComponent(computedHmac) == receivedHmac;
}

function sortObjectByKeys(o) {
  return Object.keys(o)
    .sort()
    .reduce((r, k) => ((r[k] = o[k]), r), {});
}

module.exports = {
  PaymentHandler,
  SimpleLogger,
  APIException,
  validateHMAC_SHA256,
};

File: PythonBackend-kit/paymentHandler.py
import json
import base64
import requests
import logging
from abc import ABC, abstractmethod
from typing import Any, Union
from collections.abc import MutableMapping
from logging.handlers import RotatingFileHandler
import hmac, hashlib, urllib.parse

with open('config.json', 'r') as f:
    config = json.load(f)

DEFAULT_TIMEOUT = 5 #seconds
API_VERSION = "2025-05-01"
PROJECT_VERSION = "2025-05-01"
KIT_VERSION = 'PYTHON_KIT1/' + PROJECT_VERSION

ENABLE_LOGGING = config.get('ENABLE_LOGGING')

class APIException(Exception):
    def __init__(self, http_response_code=None, status=None, error_code=None, error_message=None):
        super().__init__(error_message or error_code or "Something went wrong")
        self.http_response_code = http_response_code
        self.status = status
        self.error_code = error_code
        self.error_message = error_message

class Logger(ABC):
    @abstractmethod
    def info(self, message: Any = None) -> Union['Logger', Any]:
        pass

    @abstractmethod
    def error(self, message: Any = None) -> Union['Logger', Any]:
        pass

class SimpleLogger(Logger):
    def __init__(self, is_logger_disabled, logger: Logger = None):
        if logger is not None:
            self.logger = logger
        else:
            self.logger = logging.getLogger("Merchant")
            self.logger.setLevel(logging.INFO)

            if not self.logger.handlers:
                console_handler = logging.StreamHandler()
                console_handler.setLevel(logging.INFO)

                file_handler = RotatingFileHandler('file.log', maxBytes=2 * 1024 * 1024, backupCount=5)
                file_handler.setLevel(logging.INFO)

                formatter = logging.Formatter('Merchant : %(asctime)s - %(levelname)s - %(message)s')
                console_handler.setFormatter(formatter)
                file_handler.setFormatter(formatter)

                self.logger.addHandler(console_handler)
                self.logger.addHandler(file_handler)
        self.disable_logger = is_logger_disabled

    def info(self, message: any = None) -> Union['Logger', Any]:
        if self.disable_logger and self.logger:
            self.logger.info(message)
        return self.logger

    def error(self, message: any = None) -> Union['Logger', Any]:
        if self.disable_logger and self.logger:
            self.logger.error(message)
        return self.logger

log = SimpleLogger(ENABLE_LOGGING)

class Request:
    def __init__(self, merchant_id, base_url, auth, customer_id=None, timeout=DEFAULT_TIMEOUT, api_version=API_VERSION):
        self.merchant_id = merchant_id
        self.base_url = base_url
        self.auth = auth
        self.customer_id = customer_id
        self.timeout = timeout
        self.api_version = api_version
        self.auth_type = "BASIC"
        self.api_key = self.auth
        self.custom_headers = None
        self.query_params = None

def flatten(dictionary, parent_key=False, join_with='.'):
    items = []
    for key, value in dictionary.items():
        new_key = str(parent_key) + join_with + key if parent_key else key
        if isinstance(value, MutableMapping):
            if not value.items():
                items.append((new_key, None))
            else:
                items.extend(flatten(value, new_key, join_with).items())
        else:
            items.append((new_key, value))
    return dict(items)

def requestHeaders(request_options):
    default_headers = {
        'version': request_options.api_version,
        'User-Agent': KIT_VERSION,
        'x-merchantid': request_options.merchant_id,
        'x-customerid': request_options.customer_id
    }
    inp_headers = request_options.custom_headers
    default_headers = getAuthorizationValueWithoutPassword(request_options, default_headers)
    if inp_headers is not None:
        for k, v in default_headers.items():
            if k not in inp_headers:
                inp_headers[k] = v
        return inp_headers
    else:
        return default_headers

def getAuthorizationValueWithoutPassword(request_options, default_headers):
    if request_options.api_key is not None:
        apikey = base64.b64encode(request_options.api_key.encode('utf-8')).decode('utf-8')
        default_headers["Authorization"] = "Basic {}".format(apikey)
    return default_headers

def makeServiceCall(method, route, parameter, content_type, request_type):
    try:
        req_headers = requestHeaders(request_type)
        response = request(method, route, parameter, content_type, request_type, req_headers)
        return response
    except Exception as e:
        log.error(str(e))
        raise APIException("-1", "connection_error", "connection_error", str(e))

def request(method, route, parameters, content_type, request_type, req_headers):
    try:
        url = request_type.base_url + route
        response = makeRequest(url, method.upper(), parameters, req_headers, content_type, request_type.timeout)
        return handle_response(response)
    except IOError as err:
        raise APIException("-1", "connection_error", "connection_error", err)

def makeRequest(url, method, parameters, req_headers, content_type, timeout):
    if method == 'GET':
        return requests.get(url, headers=req_headers, params=parameters, timeout=timeout)
    elif method == 'POST':
        if content_type == "application/json":
            return requests.post(url, headers=req_headers, json=parameters, timeout=timeout)
        else:
            input_json = flatten(parameters)
            return requests.post(url, headers=req_headers, data=input_json, timeout=timeout)
    else:
        raise APIException(error_message="Method not supported")

def handle_response(response):
    if 200 <= response.status_code < 300:
        return response.json()
    else:
        if response.content:
            my_bytes_value = response.content.decode().replace("'", '"')
            if my_bytes_value:
                res = json.loads(my_bytes_value)
                status = res.get('status')
                error_code = res.get('error_code')
                error_message = res.get('error_message')
                raise APIException(response.status_code, status, error_code, error_message)
        raise APIException(response.status_code, "internal_error", "internal_error", "Something went wrong.")
class PaymentHandler:
    def __init__(self, merchant_id, base_url, auth, customer_id = None, timeout=None, api_version=None):
        self.request = Request(merchant_id, base_url, auth, customer_id, timeout, api_version)

    def validate_params(self, params):
        if not isinstance(params, dict):
            raise APIException(
                -1,
                "INVALID_PARAMS",
                "INVALID_PARAMS",
                "Params are empty or not an object"
            )
        
    def session(self, params):
        self.validate_params(params)
        path = "/session"
        method = 'POST'
        return makeServiceCall(method, path, params, "application/json", self.request)

    def order_status(self, params):
        if isinstance(params, str):
            order_id = params
        else:
            self.validate_params(params)
            order_id = params.get("order_id")

        if not order_id:
            raise APIException(
                -1, "INVALID_PARAMS", "INVALID_PARAMS",
                "order_id is missing"
            )

        path = f"/orders/{order_id}"
        method = "GET"
        return makeServiceCall(method, path, params, "application/json", self.request)

    def refund(self, params):
        if isinstance(params, str):
            order_id = params
        else:
            self.validate_params(params)
            order_id = params.get("order_id")

        if not order_id:
            raise APIException(
                -1, "INVALID_PARAMS", "INVALID_PARAMS",
                "order_id is missing"
            )

        method = 'POST'
        path = f"/orders/{order_id}/refunds"
        return makeServiceCall(method, path, params, "application/json", self.request)
class SignatureValidator:
    def __init__(self, response_key: str):
        if not isinstance(response_key, str):
            raise TypeError("Response key must be a string")
        self.response_key = response_key

    def verify_request(self, request) -> bool:
        params = request.args.to_dict(flat=True)
        return self._verify(params)
    
    def _verify(self, params: dict) -> bool:
        if not isinstance(params, dict):
            raise TypeError("Params must be a dictionary")

        received_signature = urllib.parse.unquote(params.get('signature', ''))

        filtered_params = {
            k: v for k, v in params.items()
            if k not in ['signature', 'signature_algorithm']
        }
        sorted_keys = sorted(filtered_params.keys())
        param_string = '&'.join(f"{key}={filtered_params[key]}" for key in sorted_keys)
        encoded_param_string = urllib.parse.quote(param_string, safe='')
        computed_hmac = hmac.new(
            self.response_key.encode('utf-8'),
            encoded_param_string.encode('utf-8'),
            hashlib.sha256
        ).digest()

        computed_signature = base64.b64encode(computed_hmac).decode('utf-8')
        return computed_signature == received_signature